Wrap kiosk in a cgroup with a MemoryHigh limit

yfyf opened the pull request:
Recent experiments have shown that chromium (and hence QtWebEngine) will keep consuming close to max of the available memory, making it more likely that the OOM killer will get involved due to a spike in memory usage. The spike can be caused both by chromium itself (e.g. loading a heavy page) or by routine system activities (e.g. downloading of new PlayOS bundle).

To avoid this, we start kiosk in a cgroup (via a transient systemd scope) with constrained resources. Setting MemoryHigh < MemoryMax ensures that chromium will initiate GC before the system reaches a critical state.

Unfortunately, unlike for services, systemd does not provide an `ExitType=cgroup` behaviour for scopes. Meaning it is not possible to declaratively specify that all processes in the cgroup need to be terminated when the scope exits. As a workaround, we perform manual cleanup on exit.


## Checklist

-   [ ] Changelog updated
-   [X] Code documented
-   [X] User manual updated

dividat-jgu commented on application/run-with-memory-limit.sh:
```diff
@@ -0,0 +1,77 @@
+memory_high_mb=$(( memory_high_pct * system_memory_kb / 1024 / 100))
+
+if ! [[ $memory_high_mb -gt 0 ]]; then
+    echo "Computed MemoryHigh is 0 or invalid: $memory_high_mb M"
```
Nitpick: missing B?

```suggestion
    echo "Computed MemoryHigh is 0 or invalid: $memory_high_mb MB"
```

yfyf commented on application/run-with-memory-limit.sh:
Incorporated in ee53080, but with lower-case names. Though I do find the `readonly` annotations to be purely line noise :grin:

yfyf commented:
Rebased on top of main, moved things around and added an integration test that checks that:
- the memory limit is enforced
- a JS leak leads to an OOM kill and a "clean" crash/restart

Unfortunately, the test showed that with `MemoryHigh`, the JS leak causes a freeze rather than a crash, since the kernel throttles the process and chromium gets stuck in GC. I have therefore changed to `MemoryMax`, which "at least" leads to a clean crash.

My understanding is that the "deeper" issue is that Qt ships with a Chromium version **without a memory pressure (PSI) monitor** on Linux. There is [one in the sources](https://github.com/qt/qtwebengine-chromium/blob/2001f9c899030483ffacf33603a1428abb1f83a9/chromium/components/memory_pressure/system_memory_pressure_evaluator_linux.cc), but it seems it is not used and has [been removed from Chromium](https://source.chromium.org/chromium/chromium/src/+/12703e0ece2cffc126f654ade6b642d4b21d58fd) in future versions. The [new one](https://chromium-review.googlesource.com/c/chromium/src/+/3035706) is [based on D-Bus](https://hadess.pages.freedesktop.org/low-memory-monitor/gdbus-org.freedesktop.LowMemoryMonitor.html), but I think future Qt versions will not provide integration either.

I tried to understand how chromium determines available memory and responds to cgroups without a PSI monitor, but failed. 

There is marginal use with merging this - as the tests show, this still helps to keep memory usage under control, but ultimately can lead to OOMs in unexpected memory spikes like before.

See the commits and test for more details.

cc @knuton

yfyf commented:
I am converting this to draft, because the usefulness of this is questionable. For example, the test would pass even if you set `kioskPct = 100` or you remove the memory limit wrapper entirely :joy:. And you actually get a slightly clearer crash:

```
machine # [  182.145795] xsession[1098]: <--- Last few GCs --->
machine # [  182.293862] traps: QtWebEngineProc[1098] trap int3 ip:7f46e7a6f0ad sp:7ffd0f583890 error:0 in libQt6WebEngineCore.so.6.8.1[7f46e2b29000+9ef8000]
machine # [  182.193693] xsession[1098]:  m[1:0x55edb3a43000]   161578 ms: Mark-Compact (reduce) 660.6 (661.8) -> 660.6 (661.8) MB, pooled: 0 MB, 1.63 / 0.02 ms  (+ 0.4 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 61 ms) (average mu = 0.997, current mu[1:0x55edb3a43000]   163902 ms: Mark-Compact (reduce) 720.6 (721.8) -> 720.6 (721.8) MB, pooled: 0 MB, 45.45 / 1.36 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 116 ms) (average mu = 0.988, current
machine # [  182.200912] xsession[1098]: <--- JS stacktrace --->
machine # [  182.226924] xsession[1098]: [1098:1:1126/100023.729162:ERROR:v8_initializer.cc(800)] V8 javascript OOM (Reached heap limit).
```

yfyf commented:
E2E test is failing because GH runner is out of memory, unrelated to these changes. Looking for a fix.

knuton commented:
> I vote for option 1, since the benefits seem questionable, while the complexity increase is significant. If memory is unreclaimable, kiosk/chromium _will_ OOM in both cases, since it does not know about the memory pressure.

Agreed, currently unclear this would be a net improvement.

Thanks for working through this thoroughly!
